File: leetcode/medium/264_nthUglyNumber.py
from typing import List

class Solution:
    def nthUglyNumber__(self, n: int) -> int:
        if n == 1:
            return 1

        ugly = 1
        allNumbers = set()
        allNumbers.add(ugly)

        for i in range(n):
            ugly = min(allNumbers)
            allNumbers.remove(ugly)

            allNumbers.add(ugly * 2)
            allNumbers.add(ugly * 3)
            allNumbers.add(ugly * 5)

        return ugly

    def nthUglyNumber(self, n: int) -> int:
        if n == 1:
            return 1

        allNumbers = [0] * n
        allNumbers[0] = 1
        i2, i3, i5 = 0, 0, 0
        n2, n3, n5 = 2, 3, 5

        for i in range(1, n):
            allNumbers[i] = min(n2, n3, n5)

            if allNumbers[i] == n2:
               i2 += 1
               n2 = allNumbers[i2] * 2
            if allNumbers[i] == n3:
                i3 += 1
                n3 = allNumbers[i3] * 3
            if allNumbers[i] == n5:
                i5 += 1
                n5 = allNumbers[i5] * 5

        return allNumbers[n - 1]

def test ():
    params = [
        {
            'input': 10,
            'output': 12,
        },
        {
            'input': 1,
            'output': 1,
        },
        {
            'input': 11,
            'output': 15,
        },
    ]
    solution = Solution()

    for param in params:
        result = solution.nthUglyNumber(param['input'])
        print(
            'SUCCESS' if result == param['output'] else 'ERROR',
            'input', param['input'],
            'output', param['output'],
            'result', result,
            '\n',
        )

if __name__ == '__main__':
    test()


'''
allNumbers [1, 2, 3, 4, 5, 6, 8, 9, 10, 12, 14]
ERROR input 11 output 15 result 14

'''

File: leetcode/medium/264_nthUglyNumber.ts
function nthUglyNumber_0(n: number): number {
    const multipliers = [2, 3, 5]
    const unique: Set<number> = new Set([1]);
    const arr: number[] = [1];

    let i: number = 1;

    while (i <= n) {
        if (i === n) {
            return arr[i-1];
        }

        for (const num of multipliers) {
            const m = num * arr[i-1];
            if (unique.has(m)) {
                continue;
            }

            unique.add(m);
            arr.push(m);
        }

        arr.sort((a, b) => a - b);
        i += 1;
    }


    return arr[n];
};


function nthUglyNumber(n: number): number {
    let i: number = 1;
    let id2: number = 0;
    let id3: number = 0;
    let id5: number = 0;

    const arr2: number[] = [1];
    const arr3: number[] = [1];
    const arr5: number[] = [1];

    while (i <= n) {
        const val: number = Math.min(arr2[id2], arr3[id3], arr5[id5]);
        if (i === n) {
            return val;
        }

        arr2.push(val * 2);
        arr3.push(val * 3);
        arr5.push(val * 5);

        while (arr2[id2] === val) {
            id2 += 1;
        }
        while (arr3[id3] === val) {
            id3 += 1;
        }
        while (arr5[id5] === val) {
            id5 += 1;
        }

        i += 1;
    }


    return 0;
};

/*

2 1*2=2; 2*2=4;  3*2=6;   4*2=8;  5*2=10
3 1*3=3; 2*3=6;  3*3=9;   4*3=12; 5*3=15;
5 1*5=5; 2*5=10; 3*15=15; 4*5=20; 5*4=20;


*/

const test = () => {
    const params = [
        {
            input: {
                n: 10,
            },
            output: 12,
        },
        {
            input: {
                n: 1,
            },
            output: 1,
        },
    ];

    params.forEach(({ input, output }) => {
        const { n } = input;
        const result = nthUglyNumber(n);

        console.log(
            JSON.stringify(result) === JSON.stringify(output)
                ? "SUCCESS "
                : "ERROR ",
            "input",
            JSON.stringify(input),
            "output",
            output,
            "result",
            result,
        );
    });
};

test();
